### DISCUSSION TITLE
Qual melhor forma de aprender JavaScript?

### DISCUSSION THREAD

**Discussion by adelmojnr** [frontendbr/forum#410] [Discussão]
Fico meio perdido na internet pelo fato de ter tanto conteúdo a respeito de JavaScript, ultimamente estou vendo a série de vídeos do Estevan Maito, e está me ajudando, e também lendo o MDN.   
(desculpem pela perguntar ser um pouco vaga)

Gostaria saber de vocês, qual dica podem me dar?  

Desde já, grato.

**Comment by laurindo** [comment]
Espero que ajude
https://github.com/ericdouglas/traduz-ai/blob/master/javascript/001-como-aprender-js-corretamente.md

**Comment by gabrieldarezzo** [comment]
Acredito que ter uma boa noção de DOM é muito importante, apesar que já foi a época que JS era exclusivo do uso do Browser.

@laurindo não conhecia esse repo, vou anexar no meu tutorial/exercícios (link abaixo)

Sobre o artigo, eu discordo com essa parte :
> Não tente aprender Javascript pela primeira vez através de partes não relacionadas ou relacionadas a tutoriais online de javascript

Acredito que só se tem domínio de algo praticando muito, lendo, tentando, estudando.
E de qualquer forma esse conjunto por 'pedaços' acaba virando uma rede grande.



Mantenho um repo com alguns exercícios/desafios aqui caso queira praticar:
https://github.com/gabrieldarezzo/helpjs-ravi

Porem só pelo fato de você ler o MDN, acredito que os desafios estão um nível muito básico pra ti.
Importante também entender sobre as requisições assíncronas, síncrona.


Deixo um trecho de um livro que me ajudou muito:

> Os programadores realmente bons, que unem formação técnica, criatividade, curiosidade e prazer
> parecem ser relativamente poucos. Naqueles que conheci pude perceber padrões: eles têm acesso a um
> computador, em casa ou em outro lugar; têm uma grande disposição de buscar soluções por si mesmos e, acima de
> tudo, eles programam muito. Não dependem de alguém mandá-los programar, tendo ou não suas próprias idéias.
> Para resumir, eles simplesmente praticam muito, e este é o fator maior que, na minha opinião, os torna melhores
> que os outros

'300  IDÉIAS PARA PROGRAMAR COMPUTADORES', Virgílio Vasconcelos Vilela

Acompanhando a postagem :)

**Comment by laurindo** [comment]
Esse artigo do tableless -  Tenha o Dom é bem legal, eu até usei ele para
algumas palestras.
Tem outro repositório muito bom que não achei, mas é indicações de livros
básicos, intermediários e avançados sobre JS.
Um deles é O Eloquente javascript
https://github.com/braziljs/eloquente-javascript

**Comment by thiamsantos** [comment]
Acho que a parte dois do artigo [From Zero to Front-End Hero](https://medium.freecodecamp.com/from-zero-to-front-end-hero-part-2-adfa4824da9b) apresenta um caminho bem interessante, o @LFeh fez a [tradução](http://www.felipefialho.com/blog/2016/do-zero-a-heroi-do-front-end-parte-2) desse artigo para o português. Vale a pena a leitura!

**Comment by adelmojnr** [comment]
Valeu galera

**Comment by joelbarbosa** [comment]
Tem uma série de livros nessa repo:
https://github.com/getify/You-Dont-Know-JS

**Comment by adelmojnr** [comment]
valeu @joelbarbosa xD

**Comment by mksanti** [comment]
Tenho um repo que atualizo de tempos em tempos de livros bons de JS que a comunidade traduziu para pt-br, pode ajudar. [https://github.com/mikejavier/livros-js-ptbr](https://github.com/mikejavier/livros-js-ptbr)

**Comment by jrfferreira** [comment]
@adelmojnr, cada pessoa tem formas bem distintas de aprender, alguns gostam de dedicar ao estudo de teorias, outros preferem a prática. Um livro que mescla muito bem isso, é o ["Segredos do ninja Javascript"](http://novatec.com.br/livros/ninja-javascript/) que eu particularmente gosto muito, dado que cobre muito bem a linguagem com a visão prática de se construir um framework. Também recomendaria a leitura de bons códigos, acompanhar projetos open-sources e tentar entender os dilemas que esses programadores tiveram que resolver, ou coletaneas como a ["clean-code-javascript"](https://github.com/ryanmcdermott/clean-code-javascript) que dão bons exemplos de como simplificar e esclarecer trechos de código.
Alguns styleguides, como o da [airbnb](https://github.com/airbnb/javascript) são bem completos sobre como codificar, mas são dificeis para um iniciante, pois pecam quanto a instrução ou didática.

Espero ter ajudado, um abraço!

**Comment by ingridrauany** [comment]
Eu comecei a fazer, mas ainda não terminei o curso [Pratical Javascript](https://watchandcode.com/p/practical-javascript), mas pelo início eu gostei bastante. Ele começa com códigos básicos e vai subindo o nível com as aulas.

**Comment by adelmojnr** [comment]
Valeu @jrfferreira

**Comment by adelmojnr** [comment]
Obrigado @ingridrauany

**Comment by matheuslc** [comment]
Vou contar como aprendi, que, não necessariamente é a melhor forma (e que eu ainda não esteja aprendendo).

Eu brincava de fazer sites utilizando plugins já prontos e um dia decidi que eu queria aprender à fundo como programar. Iniciei meus estudos com o livro: Professional JavaScript for Web Developes, 3rd (Nicholas C. Zakas). Esse livro me deu uma base muito sólida sobre JavaScript. Protótipos, Coletor de Lixo (Garbage Collector é mais bonito :P), Canvas, Ajax, DOM, etc. Durante os estudos com este livro, escrevia um blog dos assuntos que eu ia aprendendo para fixar o conteúdo. Mas, o mais importante ao meu ver, pra fixar o conteúdo eu fazia alguns projetos. Neste período, estava cursando Engenharia da Computação e transcrevia o que aprendia na faculdade para JavaScript (estatística, estrutura de dados etc).

Entrei num estágio e ai comecei a ser desafiado com problemas reais para serem resolvidos com JavaScript. Aprendi MUITO. Tente arrumar um estágio na área, vai te evoluir muito.

Depois de já manjar um pouco da linguagem, li mais alguns livros (Eloquent JavaScript, JavaScript for Cats,  Segredos do Ninja JavaScript, JavaScript Patterns etc), fiz vários projetinhos pessoais nos quais tentava aplicar o que aprendia e também tentava compartilhar meus conhecimentos na comunidade (e adquirir, claro).

Vá a Meetups, a galera manja de mais e te passará conhecimento e experiência que serão muito importantes (deixei um pouco de lado por causa da faculdade, mas voltarei).

No começo é meio difícil escolher com quem trabalhar, mas, além de salário etc, procure trabalhar em empresas com pessoas boas, isso vai te evoluir de mais.

Acompanhe blogs bons, ponyfoo.com, 2ality.com, davidwalsh.name, smashingmagazine.com são alguns deles. 

Se você não manja inglês, ele é fundamental, vá atrás.

Leia código de projetos open source (comece por libs pequenas e depois frameworks e libs grandes).

E não se tratando de JavaScript em si, mas aprenda os fundamentos da computação, como uma linguagem funciona, quando se vai fazer um Ajax, como o protocolo HTTP funciona etc. Tente descobrir todas as mágicas por trás das coisas.

Acho que é isso. Qualquer coisa que precisar no percurso, conte com a comunidade o/

**Comment by adelmojnr** [comment]
@matheuslc valeu cara, pra estágio está meio difícil pra mim, só tenho 16 anos, rsrsrs, mas tô atrás, tô pensando muito em comprar JavaScript o guia definitivo xD

**Comment by rogeralbinoi** [comment]
Cara, aqui tenho alguns links que podem ajudar:
https://developer.mozilla.org/pt-BR/docs/Web/JavaScript
https://www.codecademy.com/
https://www.javascript.com/

**Comment by adelmojnr** [comment]
valeu @rogeralbinoi xD

**Comment by raloliver** [comment]
Fiz uma lista de cursos, livros, ebooks, vídeos, artigos e outros que me ajudam. Não é a maneira certa, mas é o que tem funcionado pra mim.

Tenho um rotina e ainda pratico mais em casa do que no trabalho (no trabalho faço outras paradas e em casa escrevo código):

**Segunda**: Curso de JS (você escolhe um, eu to com vários e vou ouvindo o mesmo assunto de professores diferentes, isso ajuda a trazer clareza)
**Terça**: Prática Deliberada (tento pensar em como colocar em prática o que aprendi, sempre daquele jeito "se dá pra fazer com javascript, vamos fazer com javascript")
**Quarta**: Curso de AngularJS (ainda gosto do AngularJS e como tem muita demanda pra ele onde trabalho, tenho estudado.)
**Quinta**: Livro de Javascript (veja a minha lista: http://isra.me/amazon)
**Sexta**: Livro de AngularJS e algum outro curso que envolve o universo Web (git, html, css, tdd, webpack)

Sempre no meio tempo tento aprender inglês (podcast, youtube, eventos gravados, twitter e audiolivros) . Escuto muito podcast indo pro trabalho, voltando, quando vou caminhar... enfim, sempre aproveito o tempo livre pra focar em outros assuntos que permeiam a vida de um "_client side developer_" (UX, Design, Marketing, Carreira, Startups, etc).

Pra começar hoje:
1. Freecodecamp (Curso): https://www.freecodecamp.com/map (tem mais de 100 exercícios)
2. Eloquente Javascript (Ebook): https://github.com/braziljs/eloquente-javascript
3. Javascript de Alto Desempenho (Livro): http://a.co/2dLBaZx _(invista em você)_
4. Jornada do Dev Javascript (Curso): http://jornadadodev.com.br/cursos/curso-completo-de-javascript
5. Javascript Ninja: http://blog.da2k.com.br/curso-javascript-ninja/ _(invista em você)_

**P.S.: mínimo de 3 horas por noite, as vezes 4 horas. Finais de semana tento revisar o que aprendi, sempre anoto tudo.**
P.S.S.: tenho uma lista aqui de vários cursos, livros e outros que pretendo estudar no decorrer do ano, se tiver afim é só pedir que organizo e envio.

**Comment by adelmojnr** [comment]
Muito massa @raloliver, muito obrigado man, curti muito sua forma de organizar o aprendizado xD

**Comment by brunooomelo** [comment]
@raloliver muito bom, manda os cursos ou livros

**Comment by ranosi** [comment]
Algumas horas antes de tu publicar a tua rotina de estudos eu estava montando a minha @raloliver . Fiz algumas alterações com os links que tu mandou, achei bem massa o Free Code Camp. Valeu pelas infos cara.

**Comment by raloliver** [comment]
@brunooomelo 5 Livros e 5 Cursos (_acho que da pra estudar o ano inteiro_)
## LIVROS
1. [You Don't Know JS (livro)](https://github.com/cezaraugusto/You-Dont-Know-JS)*
2. [JavaScript Furtivo (livro)](https://leanpub.com/javascriptfurtivo)*
3. [Princípios: Orientação a Objetos (livro)](http://a.co/8hYmVtv)
4. [Construindo Uma Aplicação E-commerce com Mean (livro)](http://a.co/2dw6sc6)
5. [Desenvolvendo com AngularJS (livro)](http://a.co/cSiOwbg)
## CURSOS
1. [JavaScript 30 (curso)](https://javascript30.com/)* (_esse é bom pra saber das possibilidades e depois se aprofundar nos recursos utilizados em cada aula_)
2. [Pratical JavaScript (curso)](https://watchandcode.com/p/practical-javascript)* (_ainda não fiz, mas será o próximo que devo começar_)
3. [JavaScript Experts (curso)](https://www.udemy.com/curso-de-javascript-completo-e-passo-a-passo/) (_pode não ser dos melhores, mas como gosto de ter vários inputs diferentes do mesmo assunto, valeu pagar R$25_)
4. [JS com TDD na Prática (curso)](https://willianjusten.com.br/js-com-tdd-na-pratica-e-novidades/#ementa)
5. [Nanodegree Desenvolvedor Web Front-End (formação)](https://br.udacity.com/course/front-end-web-developer-nanodegree--nd001/) (_formação de um ano com diploma válido em vários países_)
## ASSINATURAS
> Sou assinante de alguns sites: **Pluralsigth**, **Alura**, **School of Net**, **TutsPlus**, **SitePoint** e **Balta.io**
> Sites que queria ter a assinatura: **Code School**, **Frontend Masters**, **Udacity** e **ES6.io**

*Conteúdo gratuito.

**Comment by diegoslva** [comment]
O tem me ajudado bastante a entender algumas coisas de JS é ter um projeto paralelo aos estudos e ir aplicando coisas que vou aprendendo no caminho. E nesse caminho sempre procuro reproduzir alguns layouts ou modulos foda com animações e tudo mais  que vejo no dribble ou qualquer outro lugar. Sendo assim vou sempre aplicando e tentando aumentar o nível de dificuldade. Por que se tal layout funciona de uma forma por que não mudar e fazer funcionar de outro jeito. A partir dai vou aplicando essas paradas no meu projeto paralelo e caindo em dificuldades mais próximos de um ambiente rotineiro.

**Comment by diegoslva** [comment]
O tem me ajudado bastante a entender algumas coisas de JS é ter um projeto paralelo aos estudos e ir aplicando coisas que vou aprendendo no caminho. E nesse caminho sempre procuro reproduzir alguns layouts ou modulos foda com animações e tudo mais  que vejo no dribble ou qualquer outro lugar. Sendo assim vou sempre aplicando e tentando aumentar o nível de dificuldade. Por que se tal layout funciona de uma forma por que não mudar e fazer funcionar de outro jeito. A partir dai vou aplicando essas paradas no meu projeto paralelo e caindo em dificuldades mais próximos de um ambiente rotineiro.

**Comment by Heggendorne** [comment]
Fala Israel, bom dia.

Primeiro, já agradeço pelo post em 2017 no github a respeito de como foi que aprendeu js, caso não lembre: https://github.com/frontendbr/forum/issues/410.
Eu sou iniciante em programação e, eu queria saber se a conciliação de Js e Angula deu certo?

Abraços.
Eduardo Heggendorne

**Comment by alexpaivadev** [comment]
> Fiz uma lista de cursos, livros, ebooks, vídeos, artigos e outros que me ajudam. Não é a maneira certa, mas é o que tem funcionado pra mim.
> 
> Tenho um rotina e ainda pratico mais em casa do que no trabalho (no trabalho faço outras paradas e em casa escrevo código):
> 
> **Segunda**: Curso de JS (você escolhe um, eu to com vários e vou ouvindo o mesmo assunto de professores diferentes, isso ajuda a trazer clareza)
> **Terça**: Prática Deliberada (tento pensar em como colocar em prática o que aprendi, sempre daquele jeito "se dá pra fazer com javascript, vamos fazer com javascript")
> **Quarta**: Curso de AngularJS (ainda gosto do AngularJS e como tem muita demanda pra ele onde trabalho, tenho estudado.)
> **Quinta**: Livro de Javascript (veja a minha lista: http://isra.me/amazon)
> **Sexta**: Livro de AngularJS e algum outro curso que envolve o universo Web (git, html, css, tdd, webpack)
> 
> Sempre no meio tempo tento aprender inglês (podcast, youtube, eventos gravados, twitter e audiolivros) . Escuto muito podcast indo pro trabalho, voltando, quando vou caminhar... enfim, sempre aproveito o tempo livre pra focar em outros assuntos que permeiam a vida de um "_client side developer_" (UX, Design, Marketing, Carreira, Startups, etc).
> 
> Pra começar hoje:
> 1. Freecodecamp (Curso): https://www.freecodecamp.com/map (tem mais de 100 exercícios)
> 2. Eloquente Javascript (Ebook): https://github.com/braziljs/eloquente-javascript
> 3. Javascript de Alto Desempenho (Livro): http://a.co/2dLBaZx _(invista em você)_
> 4. Jornada do Dev Javascript (Curso): http://jornadadodev.com.br/cursos/curso-completo-de-javascript
> 5. Javascript Ninja: http://blog.da2k.com.br/curso-javascript-ninja/ _(invista em você)_
> 
> **P.S.: mínimo de 3 horas por noite, as vezes 4 horas. Finais de semana tento revisar o que aprendi, sempre anoto tudo.**
> P.S.S.: tenho uma lista aqui de vários cursos, livros e outros que pretendo estudar no decorrer do ano, se tiver afim é só pedir que organizo e envio.

Amigo, pode me mandar a lista de conteúdo. alexpaiva.dev@gmail.com
Obrigado desde já

**Comment by raloliver** [comment]
> 
> 
> Fala Israel, bom dia.
> 
> Primeiro, já agradeço pelo post em 2017 no github a respeito de como foi que aprendeu js, caso não lembre: #410.
> Eu sou iniciante em programação e, eu queria saber se a conciliação de Js e Angula deu certo?
> 
> Abraços.
> Eduardo Heggendorne

Eu me perdi nas notificações e só agora li isso... e sim, deu certo.

**Comment by raloliver** [comment]
> 
> 
> > Fiz uma lista de cursos, livros, ebooks, vídeos, artigos e outros que me ajudam. Não é a maneira certa, mas é o que tem funcionado pra mim.
> > Tenho um rotina e ainda pratico mais em casa do que no trabalho (no trabalho faço outras paradas e em casa escrevo código):
> > **Segunda**: Curso de JS (você escolhe um, eu to com vários e vou ouvindo o mesmo assunto de professores diferentes, isso ajuda a trazer clareza)
> > **Terça**: Prática Deliberada (tento pensar em como colocar em prática o que aprendi, sempre daquele jeito "se dá pra fazer com javascript, vamos fazer com javascript")
> > **Quarta**: Curso de AngularJS (ainda gosto do AngularJS e como tem muita demanda pra ele onde trabalho, tenho estudado.)
> > **Quinta**: Livro de Javascript (veja a minha lista: http://isra.me/amazon)
> > **Sexta**: Livro de AngularJS e algum outro curso que envolve o universo Web (git, html, css, tdd, webpack)
> > Sempre no meio tempo tento aprender inglês (podcast, youtube, eventos gravados, twitter e audiolivros) . Escuto muito podcast indo pro trabalho, voltando, quando vou caminhar... enfim, sempre aproveito o tempo livre pra focar em outros assuntos que permeiam a vida de um "_client side developer_" (UX, Design, Marketing, Carreira, Startups, etc).
> > Pra começar hoje:
> > 
> > 1. Freecodecamp (Curso): https://www.freecodecamp.com/map (tem mais de 100 exercícios)
> > 2. Eloquente Javascript (Ebook): https://github.com/braziljs/eloquente-javascript
> > 3. Javascript de Alto Desempenho (Livro): http://a.co/2dLBaZx _(invista em você)_
> > 4. Jornada do Dev Javascript (Curso): http://jornadadodev.com.br/cursos/curso-completo-de-javascript
> > 5. Javascript Ninja: http://blog.da2k.com.br/curso-javascript-ninja/ _(invista em você)_
> > 
> > **P.S.: mínimo de 3 horas por noite, as vezes 4 horas. Finais de semana tento revisar o que aprendi, sempre anoto tudo.**
> > P.S.S.: tenho uma lista aqui de vários cursos, livros e outros que pretendo estudar no decorrer do ano, se tiver afim é só pedir que organizo e envio.
> 
> Amigo, pode me mandar a lista de conteúdo. [alexpaiva.dev@gmail.com](mailto:alexpaiva.dev@gmail.com)
> Obrigado desde já

WOW! Muita coisa aconteceu desde que escrevi isso... e todas as anotações que tenho aqui são antigas e alguns links já não funcionam mais... o que funcionou pra mim foi escolher bem os professores (as vezes assino um site so por causa de um professor) praticar mais e diariamente, aprender com professores diferentes e pegar um livro + um ebook para exercícios diários + desafios na internet para resolver e bater a cabeça até entender, além, de cursos, youtube e tutoriais... deixei aqui alguns links de sites pra praticar: https://gist.github.com/raloliver/6a36bd26e4282ffcfbf19be2d4c490ee

**Comment by allanppss** [comment]
https://tutors.a7apps.com/cursos-gratuitos/curso-de-javascript/
